### PR TITLE
feat: support empty labels and label matchers

### PR DESCRIPTION
**Pull request by thisisnithin**
## Motivation and Context

image
image

## TODO

- [ ] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

File: controlplane/src/core/repositories/SubgraphRepository.ts [hunk 1]
```diff
@@ -237,8 +238,8 @@ export class SubgraphRepository {
         labelChanged = hasLabelsChanged(subgraph.labels, data.labels);
       }
 
-      if (labelChanged && data.labels) {
-        const newLabels = normalizeLabels(data.labels);
+      if (labelChanged || data.unsetLabels) {
+        const newLabels = normalizeLabels(data.unsetLabels ? [] : data.labels);
```
Comment (StarpTech) [inline controlplane/src/core/repositories/SubgraphRepository.ts:242]: I'd suggest this

```ts
const newLabels = data.unsetLabels ? [] : normalizeLabels(data.labels);
```

File: controlplane/src/core/repositories/SubgraphRepository.ts [hunk 2]
```diff
@@ -874,6 +875,11 @@ export class SubgraphRepository {
       conditions.push(sql.raw(`labels && '{${labelsSQL}}'`));
     }
 
+    // Only get subgraphs that do not have any labels if the label matchers are empty.
+    if (data.labelMatchers.length === 0) {
+      conditions.push(eq(targets.labels, []));
+    }
+
     const subgraphs = await this.db
```
Comment (StarpTech) [inline controlplane/src/core/repositories/SubgraphRepository.ts:883]: We forget to use `tx` here.

File: controlplane/test/label.test.ts [hunk 3]
```diff
@@ -248,4 +248,299 @@ describe('Labels', (ctx) => {
...
+    });
+    expect(graph1AfterUnset.response?.code).toBe(EnumStatusCode.OK);
+    expect(graph1AfterUnset.graph?.labelMatchers.length).toBe(0);
+    expect(graph1AfterUnset.subgraphs.length).toBe(1);
+    expect(graph1AfterUnset.subgraphs[0].name).toBe(subgraph2Name);
+
+    await server.close();
+  });
+
+  test('Unset and set subgraph label', async (testContext) => {
```
Comment (StarpTech) [inline controlplane/test/label.test.ts:453]: Could you be a bit more descriptive here? What use cases does it test? I see the following cases:

A User creates a fed graph or subgraph first with empty labels/matchers
A User adds labels/matchers afterwards
A User unsets labels/matchers from an existing graph

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: cli/src/commands/subgraph/commands/publish.ts [hunk 4]
```diff
@@ -29,6 +29,10 @@ export default (opts: BaseCommandOptions) => {
     'The labels to apply to the subgraph. The labels are passed in the format <key>=<value> <key>=<value>. Required to create the subgraph.',
     [],
   );
+  command.option(
+    '--unset-labels',
```
Comment (StarpTech) [inline cli/src/commands/subgraph/commands/publish.ts:33]: :+1:

File: controlplane/src/core/repositories/FederatedGraphRepository.ts [hunk 5]
```diff
@@ -162,21 +164,25 @@ export class FederatedGraphRepository {
...
-              targetId: federatedGraph.targetId,
-              labelMatcher: labelMatcher.split(','),
-            })),
-          )
-          .execute();
+        if (labelMatchers.length > 0) {
+          await tx
+            .insert(schema.targetLabelMatchers)
+            .values(
+              labelMatchers.map((labelMatcher) => ({
```
Comment (StarpTech) [inline controlplane/src/core/repositories/FederatedGraphRepository.ts:179]: Don't we need to call `normalizeLabelMatchers` here?

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: controlplane/test/label.test.ts (hunk 3 above)
Comment (thisisnithin) [reply controlplane/test/label.test.ts:467]: It is difficult to describe user flow in a sentence. I have added comments. I think I have covered all cases

File: controlplane/src/core/repositories/FederatedGraphRepository.ts (hunk 5 above)
Comment (thisisnithin) [reply controlplane/src/core/repositories/FederatedGraphRepository.ts:179]: It is called

File: controlplane/src/core/repositories/SubgraphRepository.ts (hunk 1 above)
Comment (thisisnithin) [reply controlplane/src/core/repositories/SubgraphRepository.ts]: Done

File: controlplane/src/core/repositories/SubgraphRepository.ts (hunk 2 above)
Comment (thisisnithin) [reply controlplane/src/core/repositories/SubgraphRepository.ts:883]: The repository is created in a transaction, all good here

Comment (StarpTech) [reply controlplane/src/core/repositories/SubgraphRepository.ts:883]: ok

File: controlplane/test/label.test.ts (hunk 3 above)
Comment (StarpTech) [reply controlplane/test/label.test.ts:467]: You can try to describe the test expectation e.g. `Should compose correctly when removing all labels from a subgraph.` If you end up in a very long sentences it is a good sign to split up your test.

Comment (thisisnithin) [reply controlplane/test/label.test.ts]: Improved them

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**